search/LocalSearch_First_NQueensProblem.cpp

```diff
@@ -0,0 +1,173 @@
+ */
+
+/*
+* The LS algorithm is the basis of many methods used for optimization problems. 
+* Your process can be seen as iterative starting with an initial solution and 
+* improving it by making local modifications.  Its system is based on the search 
+* for neighborhoods that obtain a better solution.
+* The neighborhood search and pivot rule used for replacement are usually:
+*  - Select the best neighbor of all (best-improvement rule).
+*  - Select the first neighbor that improves the solution (first-improvement rule).
+* The rule applied in this file is first.improvement.
+*/
+
+/*
+* Problem solving the algorithm
+* The eight queens problem is a very popular strategy that consists of placing eight 
+* queens within a chessboard without threatening them. Finding the solution for different 
+* dimensions is the same as finding the solution for a minimal problem.
+*/
+
+#include "LocalSearchFirst.h"
+#include <iostream>
+#include <vector>
+#include <algorithm>
+#include <list>
+#include <time.h>
+#define NREINAS 8
+#define t_neighbor 15
+using namespace std;
+
+vector<vector<int>> neighbour_sys;
+
+void print_vector(vector<int> s) {
+	for (int x : s) { printf("%d ", x); }
+	printf("\n");
+}
+
+vector<int> initial_solution() {
+	vector<int> solution;
+	srand(time(NULL));
+	for (size_t i = 0; i < NREINAS; i++)
+	{
+		solution.push_back(rand() % NREINAS + 1);
+	}
+	return solution;
+}
+
+int cost(vector<int> queens_sol) {
+	int h = 0;
+	vector<int> aux_sol;
+	//sum column
+	for (size_t i = 0; i < NREINAS; i++)
+	{
+		for (size_t j = i + 1; j < NREINAS; j++)
+		{
+			if (queens_sol[i] == queens_sol[j])
+			{
+				h++;
+				break;
+			}
+		}
+	}
+	//sum diagonal
+	for (size_t i = 0; i < NREINAS; i++)
+	{
+		aux_sol.push_back(queens_sol[i] - (i + 1));
+	}
+	for (size_t i = 0; i < NREINAS; i++)
+	{
+		for (size_t j = i + 1; j < NREINAS; j++)
+		{
+			if (aux_sol[i] == aux_sol[j])
+			{
+				h++;
+				break;
+			}
+		}
+	}
+	aux_sol.clear();
+	//sum inverse diagonal
+	for (size_t i = 0; i < NREINAS; i++)
+	{
+		aux_sol.push_back(queens_sol[i] + (i + 1));
+	}
+	for (size_t i = 0; i < NREINAS; i++)
+	{
+		for (size_t j = i + 1; j < NREINAS; j++)
+		{
+			if (aux_sol[i] == aux_sol[j])
+			{
+				h++;
+				break;
+			}
+		}
+	}
+
+	return h;
+}
+
+vector<int> neighbour(vector<int> init_sol) {
+	srand(time(NULL));
+	int i = rand() % NREINAS + 1;
+	srand(time(NULL));
+	int mov = (rand() % NREINAS + 1) - i;
+	mov = (mov < 0) ? mov *= -1 : mov;
+	mov = (mov == 0) ? mov = 1 : mov;
+	init_sol[i - 1] = mov;
+	return init_sol;
+}
+
+void print_solution(vector<int> s) {
+	char table[NREINAS][NREINAS] = {};
+	for (size_t i = 0; i < NREINAS; i++)
+	{
+		table[i][s[i] - 1] = 'Q';
+	}
+	for (size_t i = 0; i < NREINAS; i++)
+	{
+		for (size_t j = 0; j < NREINAS; j++)
+		{
+			if (table[i][j] != 'Q')
+			{
+				printf(" _ ");
+			}
+			else { printf(" %c ", table[i][j]); }
+		}
+		printf("\n");
+	}
+	printf("\nSolucion: ");

```

Comment:
Avoid using C functions in C++ programs.
If this is a C program, the https://github.com/TheAlgorithms/C repository is the appropiate place. 🙂

Please check other parts of the code using `printf()`.